File: packages/api/src/models/Webhook.ts
/**
 * Webhook Model — HTTP callbacks for external integrations
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';

export type WebhookEvent =
  | 'annotation.created'
  | 'annotation.updated'
  | 'annotation.deleted'
  | 'book.started'
  | 'book.completed'
  | 'book.updated'
  | 'session.started'
  | 'session.ended'
  | 'flashcard.created'
  | 'flashcard.reviewed';

const VALID_EVENTS: WebhookEvent[] = [
  'annotation.created',
  'annotation.updated',
  'annotation.deleted',
  'book.started',
  'book.completed',
  'book.updated',
  'session.started',
  'session.ended',
  'flashcard.created',
  'flashcard.reviewed',
];

export function isValidWebhookEvent(event: string): event is WebhookEvent {
  return VALID_EVENTS.includes(event as WebhookEvent);
}

export function getValidWebhookEvents(): WebhookEvent[] {
  return [...VALID_EVENTS];
}

interface WebhookAttributes {
  id: string;
  userId: string;
  url: string;
  events: WebhookEvent[];
  secret: string;
  isActive: boolean;
  lastDeliveryAt: Date | null;
  lastDeliveryStatus: number | null;
  failureCount: number;
  createdAt: Date;
  updatedAt: Date;
}

interface WebhookCreationAttributes extends Optional<WebhookAttributes, 'id' | 'isActive' | 'lastDeliveryAt' | 'lastDeliveryStatus' | 'failureCount' | 'createdAt' | 'updatedAt'> {}

export class Webhook extends Model<WebhookAttributes, WebhookCreationAttributes> implements WebhookAttributes {
  public id!: string;
  public userId!: string;
  public url!: string;
  public events!: WebhookEvent[];
  public secret!: string;
  public isActive!: boolean;
  public lastDeliveryAt!: Date | null;
  public lastDeliveryStatus!: number | null;
  public failureCount!: number;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

Webhook.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'users', key: 'id' },
      onDelete: 'CASCADE',
    },
    url: {
      type: DataTypes.STRING(2048),
      allowNull: false,
      validate: { isUrl: true },
    },
    events: {
      type: DataTypes.JSONB,
      allowNull: false,
      defaultValue: [],
      validate: {
        isValidEventArray(value: unknown[]) {
          if (!Array.isArray(value)) throw new Error('events must be an array');
          for (const ev of value) {
            if (!VALID_EVENTS.includes(ev as WebhookEvent)) {
              throw new Error(`Invalid event: ${ev}`);
            }
          }
        },
      },
    },
    secret: {
      type: DataTypes.STRING(64),
      allowNull: false,
    },
    isActive: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
    },
    lastDeliveryAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    lastDeliveryStatus: {
      type: DataTypes.INTEGER,
      allowNull: true,
    },
    failureCount: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    createdAt: { type: DataTypes.DATE, allowNull: false },
    updatedAt: { type: DataTypes.DATE, allowNull: false },
  },
  {
    sequelize,
    tableName: 'webhooks',
    timestamps: true,
    indexes: [
      { fields: ['user_id'] },
      { fields: ['is_active'] },
    ],
  }
);

File: packages/server/app/models/webhook.py
"""Webhook and delivery log models."""

from datetime import datetime
from typing import TYPE_CHECKING, Optional
from uuid import UUID

from sqlalchemy import Boolean, ForeignKey, Index, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.db import Base

if TYPE_CHECKING:
    from app.models.user import User

VALID_WEBHOOK_EVENTS = (
    'annotation.created',
    'annotation.updated',
    'annotation.deleted',
    'book.started',
    'book.completed',
    'book.updated',
    'session.started',
    'session.ended',
    'flashcard.created',
    'flashcard.reviewed',
)


class Webhook(Base):
    __tablename__ = 'webhooks'

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        server_default=text('gen_random_uuid()'),
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    url: Mapped[str] = mapped_column(String(2048), nullable=False)
    events: Mapped[list] = mapped_column(JSONB, default=[])
    secret: Mapped[str] = mapped_column(String(64), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    last_delivery_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    last_delivery_status: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    failure_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(),
        onupdate=func.now(),
    )

    user: Mapped['User'] = relationship('User', back_populates='webhooks')
    delivery_logs: Mapped[list['WebhookDeliveryLog']] = relationship(
        'WebhookDeliveryLog',
        back_populates='webhook',
        cascade='all, delete-orphan',
    )


class WebhookDeliveryLog(Base):
    __tablename__ = 'webhook_delivery_logs'
    __table_args__ = (
        Index('ix_wh_delivery_created', 'created_at'),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        server_default=text('gen_random_uuid()'),
    )
    webhook_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('webhooks.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    event: Mapped[str] = mapped_column(String(50), nullable=False)
    url: Mapped[str] = mapped_column(String(2048), nullable=False)
    status_code: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    duration_ms: Mapped[int] = mapped_column(Integer, default=0)
    error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())

    webhook: Mapped['Webhook'] = relationship(
        'Webhook',
        back_populates='delivery_logs',
    )
    user: Mapped['User'] = relationship('User')
